src/main/groovy/wooga/gradle/build/unity/internal/BuildEngineArgs.groovy

```diff
@@ -1,58 +1,62 @@
-package wooga.gradle.build.unity.internal
-
-import org.gradle.api.provider.ListProperty
-import org.gradle.api.provider.Provider
-import org.gradle.api.provider.ProviderFactory
-
-class BuildEngineArgs {
-
-    private final ProviderFactory providers;
-    final Provider<String> method;
-    private Map<String, BuildEngineArg> args
-    private List<BuildEngineRawArg> rawArgs
-    private Provider<Map<String,?>> environment;
-
-    BuildEngineArgs(ProviderFactory providers, Provider<String> method) {
-        this.args = new HashMap<>()
-        this.rawArgs = new ArrayList<>()
-        this.providers = providers
-        this.method = method
-        this.environment = providers.provider{ new HashMap<String, ?>() };
-    }
-
-    void addArg(String key, Provider<?> rawValueProvider) {
-        args.put(key, new BuildEngineArg(key, rawValueProvider))
-    }
-
-    void addRawArgs(Provider<List<?>> rawArgsProvider) {
-        rawArgs.add(new BuildEngineRawArg(rawArgsProvider))
-    }
-
-    void addEnvs(Provider<Map<String, ?>> envsProvider) {
-        this.environment = this.environment.map {
-            it.putAll(envsProvider.get())
-            return it
-        }
-    }
-
-    Provider<Map<String, ?>> getEnvironment() {
-        return environment
-    }
-
-    Provider<List<String>> getArgsProviders() {
-        return providers.provider {
-            def allArgs = new ArrayList<String>()
-            args.values().each {
-                if(it.argStringProvider.present) {
-                    allArgs.addAll(it.argStringProvider.get())
-                }
-            }
-            rawArgs.each {
-                if(it.argStringProvider.present) {
-                    allArgs.addAll(it.argStringProvider.get())
-                }
-            }
-            return allArgs
-        }
-    }
-}
+package wooga.gradle.build.unity.internal
+
+
+import org.gradle.api.provider.Provider
+import org.gradle.api.provider.ProviderFactory
+
+class BuildEngineArgs {
+
+    private final ProviderFactory providers
+    final Provider<String> method
+    private Map<String, BuildEngineArg> args
+    private List<BuildEngineRawArg> rawArgs
+    private Provider<Map<String, ?>> environment
+
+    BuildEngineArgs(ProviderFactory providers, Provider<String> method) {
+        this.args = new HashMap<>()
+        this.rawArgs = new ArrayList<>()
+        this.providers = providers
+        this.method = method
+        this.environment = providers.provider { new HashMap<String, ?>() }
+    }
+
+    void addArg(String key, Provider<String> flag, Provider<Object> value) {
+        args[key] = new BuildEngineArg(flag, value)
+    }
+
+    void addArg(String key, Provider<Object> value) {
+        addArg(key, providers.provider({ key }), value)
+    }
```

Comment:
In the context of arguments, I am confused as to the difference between a key and a flag. I would think the key IS the flag that is submitted through the CLI.

Reply:
It is normally. To make sure that each option is only added once, last one wins. But I think a Provider<?> object is not hashable. Thats why I did a split in this API. I'm not 100% happy with the BuildEngineArgs to begin with.